### PR TITLE
Fixed #25005 - Made date and time fields with auto_now/auto_now_add use effective default

### PR DESCRIPTION
**Pull request by yakky**
https://code.djangoproject.com/ticket/25005

Also fixed autodetector in migrations to skip promt for default for these fields.

Rebase / fix of #4902

### CODE REVIEW

File: django/db/backends/base/schema.py [hunk 1]
```diff
@@ -201,6 +202,13 @@ def effective_default(self, field):
                 default = six.binary_type()
             else:
                 default = six.text_type()
+        elif getattr(field, 'auto_now', False) or getattr(field, 'auto_now_add', False):
+            default = datetime.now()
```
Comment (charettes) [inline django/db/backends/base/schema.py]: `timezone.now()` should be used for `DateTimeField`.

Comment (yakky) [inline django/db/backends/base/schema.py]: According to the discussion here https://github.com/django/django/pull/4902 for auto_now / auto_now_add no tz should be used for DateField / TimeField (see https://github.com/django/django/pull/4902#discussion_r54287837)

Comment (charettes) [inline django/db/backends/base/schema.py]: Agreed but in the case of `DateTimeField` we should use `timezone.now()` instead just like [`DateTimeField.pre_save` does](https://github.com/django/django/blob/master/django/db/models/fields/__init__.py#L1449-L1452).

Comment (charettes) [inline django/db/backends/base/schema.py]: If we don't and `USE_TZ=True` the default datetime used on column creation will be in the system timezone instead of UTC like Django assumes datetime are stored in the database.

Comment (yakky) [inline django/db/backends/base/schema.py]: Fixed

File: tests/schema/tests.py [hunk 2]
```diff
@@ -3,6 +3,8 @@
 import unittest
 from copy import copy
 
+import mock
```
Comment (timgraham) [inline tests/schema/tests.py]: `from django.test import mock` -- this uses the package on Python 2 or the built-in version on Python 3.

File: tests/requirements/base.txt [hunk 3]
```diff
@@ -10,3 +10,4 @@ pytz > dev
 selenium
 sqlparse
 tblib
+mock
```
Comment (timgraham) [inline tests/requirements/base.txt]: chop (this is in `requirements/py2.txt`)

File: django/db/migrations/autodetector.py [hunk 4]
```diff
@@ -799,9 +799,11 @@ def _generate_added_field(self, app_label, model_name, field_name):
         # You can't just add NOT NULL fields with no default or fields
         # which don't allow empty strings as default.
         preserve_default = True
+        time_fields = (models.DateField, models.DateTimeField, models.TimeField)
         if (not field.null and not field.has_default() and
                 not isinstance(field, models.ManyToManyField) and
-                not (field.blank and field.empty_strings_allowed)):
+                not (field.blank and field.empty_strings_allowed) and
+                not (isinstance(field, time_fields) and (field.auto_now or field.auto_now_add))):
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: I wonder if we need the `isinstance` checks? Presumably some third-party fields might implement those options too. Maybe something like `getattr(field, auto_now, False) or ...` would be fine?

File: tests/schema/tests.py [hunk 5]
```diff
@@ -1819,3 +1822,67 @@ def test_alter_field_add_db_index_to_charfield_with_unique(self):
...
+        self.assertNotIn("date_time_of_birth_auto_now_add", columns)
+        self.assertNotIn("time_of_birth_auto_now", columns)
+        self.assertNotIn("time_of_birth_auto_now_add", columns)
+        # Create a row
+        Author.objects.create(name='Anonymous1')
+        # Add new date and datetime fields to ensure default will be
+        # used from effective_default
+        date_of_birth_auto_now = DateField(auto_now=True)
+        date_of_birth_auto_now.set_attributes_from_name('date_of_birth_auto_now')
+
```
Comment (MarkusH) [inline tests/schema/tests.py]: I'd chop every other empty line and group the `auto_now` and `auto_now_add`, but that's just cosmetics.

File: tests/schema/tests.py [hunk 6]
```diff
@@ -1819,3 +1822,67 @@ def test_alter_field_add_db_index_to_charfield_with_unique(self):
...
+        time_of_birth_auto_now = TimeField(auto_now=True)
+        time_of_birth_auto_now.set_attributes_from_name('time_of_birth_auto_now')
+
+        time_of_birth_auto_now_add = TimeField(auto_now_add=True)
+        time_of_birth_auto_now_add.set_attributes_from_name('time_of_birth_auto_now_add')
+
+        # Ensure fields was added with the right defaults
+        assertDefault(date_of_birth_auto_now, 'date_of_birth_auto_now', now.date())
+
```
Comment (MarkusH) [inline tests/schema/tests.py]: Same as above.

File: django/db/migrations/autodetector.py (hunk 4 above)
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: Since we eventually want to get rid of `auto_now` and `auto_now_add`, I'd stick to the `isinstance` check. 3rd party fields can instead inherit from the date and time fields Django provides.

File: docs/releases/1.10.txt [hunk 7]
```diff
@@ -339,6 +339,11 @@ Models
 * :meth:`QuerySet.bulk_create() <django.db.models.query.QuerySet.bulk_create>`
   sets the primary key on objects when using PostgreSQL.
 
+* `auto_now`, `auto_now_add` for
+  :class:`~django.db.models.DateField`, :class:`~django.db.models.TimeField`,
```
Comment (MarkusH) [inline docs/releases/1.10.txt]: Put the `DateTimeField` before the `TimeField`, please

**Comment by MarkusH** [comment]
I'm not sure if it's correct to automatically populate fields with `auto_now_add` with the current time when e.g. adding a field. The records in the database are already there and are being updated, not changed.

File: tests/schema/tests.py (hunk 2 above)
Comment (yakky) [inline tests/schema/tests.py]: Sorry, missed that.
Fixed

File: tests/requirements/base.txt (hunk 3 above)
Comment (yakky) [inline tests/requirements/base.txt]: Fixed

File: docs/releases/1.10.txt (hunk 7 above)
Comment (yakky) [inline docs/releases/1.10.txt]: Done

File: tests/schema/tests.py (hunk 5 above)
Comment (yakky) [inline tests/schema/tests.py]: Keeping them grouped by attribute looks more legible to me, but I will group them if you think it's better

**Comment by yakky** [comment]
Currently looks like keeping ``auto_now_add`` out is not handled well. I'll give a second look at this shortly

File: tests/schema/tests.py (hunk 5 above)
Comment (MarkusH) [inline tests/schema/tests.py]: I don't mind either way, it just took a second to spot the differences between the groups.

**Comment by yakky** [comment]
@MarkusH there are two factors at play here:
* not all database handle all the fields with a timestamp datatype (see ``ALTER TABLE "schema_author" ADD COLUMN "date_of_birth_auto_now_add" date DEFAULT %s NOT NULL`` generated by postgres for ``DateField``)
* when adding fields to an existing table, default must be provided, because database won't provide it

**Comment by MarkusH** [comment]
> * when adding fields to an existing table, default must be provided, because database won't provide it

Sure, which is why I was thinking about removing the `auto_now_add` field entirely from this PR.

**Comment by yakky** [comment]
After discussing this with @MarkusH we came up with this solution:
- ``auto_now`` does not ask for default and just use ``timezone.now`` as default
- ``auto_now_add`` prompt for a default but allows to accept the provided default (which is ``timezone.now``)

This PR also adds the ``default`` argument to ``InteractiveMigrationQuestioner._ask_default``

As discussed with @timgraham I made one separate commit for ``_ask_default`` change

**Comment by yakky** [comment]
because it's so hard to mock stdin

**Comment by yakky** [comment]
@timgraham @MarkusH I completed the changes we discussed, let me know if further changes are needed

File: django/db/migrations/questioner.py [hunk 8]
```diff
@@ -101,17 +101,35 @@ def _choice_input(self, question, choices):
                 pass
             result = input("Please select a valid option: ")
 
-    def _ask_default(self):
+    def _ask_default(self, default=''):
+        """
+        This method prompts the user for a default value to assign to the
+        ``default`` attribute.
+
+        ``default`` argument allows to provide a custom default value
```
Comment (MarkusH) [inline django/db/migrations/questioner.py]:
```
The ``default`` ...
```

File: django/db/migrations/questioner.py [hunk 9]
```diff
@@ -101,17 +101,35 @@ def _choice_input(self, question, choices):
...
+        (as string) which will be shown to the user and eventually used
+        as return value if the user does not provide any input.
```
Comment (MarkusH) [inline django/db/migrations/questioner.py]: "any other input"

File: django/db/migrations/questioner.py [hunk 10]
```diff
@@ -101,17 +101,35 @@ def _choice_input(self, question, choices):
...
+        """
         print("Please enter the default value now, as valid Python")
+        if default:
+            print("The default '{0}' is available, you can accept it by hitting 'Enter' "
```
Comment (MarkusH) [inline django/db/migrations/questioner.py]: s/hitting/pressing/ ?

File: django/db/migrations/questioner.py [hunk 11]
```diff
@@ -202,6 +207,24 @@ def ask_merge(self, app_label):
             False,
         )
 
+    def ask_auto_now_add_addition(self, field_name, model_name):
+        "Adding a auto_now_add field to a model"
+        if not self.dry_run:
+            choice = self._choice_input(
+                "You are trying to add field '%s' with 'auto_now_add' attribute to %s without a "
```
Comment (MarkusH) [inline django/db/migrations/questioner.py]: A few grammar fixes:

`You are trying to add the field '%s' with 'auto_now_add=True' to %s without a default; the database needs to set a default value for existing fields.`

File: docs/releases/1.10.txt [hunk 12]
```diff
@@ -361,6 +361,15 @@ Models
 
 * A proxy model may now inherit multiple proxy models that share a common
   non-abstract parent class.
+* `auto_now` for
+  :class:`~django.db.models.DateField`, :class:`~django.db.models.DateTimeField`,
+  :class:`~django.db.models.TimeField` does not trigger anymore request
+  for `default` value when generating migrations.
+
+* The migration question for a default value when adding `auto_now_add` for
```
Comment (MarkusH) [inline docs/releases/1.10.txt]: "The migration framework will ask for a default ..."

File: docs/releases/1.10.txt [hunk 13]
```diff
@@ -361,6 +361,15 @@ Models
...
+  :class:`~django.db.models.DateField`, :class:`~django.db.models.DateTimeField`,
+  :class:`~django.db.models.TimeField` provides now the option to set it at time
```
Comment (MarkusH) [inline docs/releases/1.10.txt]: "... TimeField unless an explicit default is given."

File: tests/migrations/test_auto_now_add/__init__.py [hunk 14]
```diff
@@ -0,0 +1,2 @@
+# -*- coding: utf-8 -*-
```
Comment (MarkusH) [inline tests/migrations/test_auto_now_add/__init__.py]: remove the lines in this file, please.

File: tests/migrations/test_autodetector.py [hunk 15]
```diff
@@ -633,6 +645,36 @@ def test_add_field(self):
...
+        changes = autodetector._detect_changes()
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, 'testapp', 1)
+        self.assertOperationTypes(changes, 'testapp', 0, ["AddField", "AddField", "AddField"])
+        self.assertOperationFieldAttributes(changes, "testapp", 0, 0, auto_now=True)
+        self.assertOperationFieldAttributes(changes, "testapp", 0, 1, auto_now=True)
+        self.assertOperationFieldAttributes(changes, "testapp", 0, 2, auto_now=True)
+
+    @mock.patch('django.db.migrations.questioner.MigrationQuestioner.ask_not_null_addition',
+                side_effect=AssertionError("Should not have prompted for not null addition"))
```
Comment (MarkusH) [inline tests/migrations/test_autodetector.py]: Can you also check for `questioner.ask_auto_now_add_addition` to be called 3 times, please or is this something we don't do in here but in commands.

**Comment by MarkusH** [comment]
Alright, looks good to me. I let @timgraham have another run through the docs before merging. Thanks!

File: tests/migrations/test_autodetector.py (hunk 15 above)
Comment (yakky) [inline tests/migrations/test_autodetector.py]: I added a new test to check the number of calls https://github.com/django/django/pull/6300/files#diff-c11e6432df7086eda3dfb9ab8e5b2839R691

**Comment by yakky** [comment]
@timgraham any more work is needed on this?

File: docs/ref/models/fields.txt [hunk 16]
```diff
@@ -521,12 +521,22 @@ optional arguments:
...
+        having a timestamp for the record **creation** while applying it's
+        just an update.
+
 The default form widget for this field is a
 :class:`~django.forms.TextInput`. The admin adds a JavaScript calendar,
 and a shortcut for "Today". Includes an additional ``invalid_date`` error
 message key.
 
-The options ``auto_now_add``, ``auto_now``, and ``default`` are mutually exclusive.
+The options ``auto_now_add``, ``auto_now``, and ``default`` are mutually
```
Comment (timgraham) [inline docs/ref/models/fields.txt]: Try to avoid making unrelated cosmetic cleanups in a patch like this because it confuses things and makes git blame less helpful. I'm just going to revert the changes to this file. The versionchanged annotation doesn't seem necessary to me.

File: docs/releases/1.10.txt [hunk 17]
```diff
@@ -365,6 +365,16 @@ Models
 
 * A proxy model may now inherit multiple proxy models that share a common
   non-abstract parent class.
+* `auto_now` for
```
Comment (timgraham) [inline docs/releases/1.10.txt]: Going to omit the release notes as these points seem more like bug fixes rather than "Minor features".

File: tests/schema/tests.py [hunk 18]
```diff
@@ -1832,3 +1833,67 @@ def test_alter_pk_with_self_referential_field(self):
...
+        date_time_of_birth_auto_now_add.set_attributes_from_name('date_time_of_birth_auto_now_add')
+
+        time_of_birth_auto_now = TimeField(auto_now=True)
+        time_of_birth_auto_now.set_attributes_from_name('time_of_birth_auto_now')
+
+        time_of_birth_auto_now_add = TimeField(auto_now_add=True)
+        time_of_birth_auto_now_add.set_attributes_from_name('time_of_birth_auto_now_add')
+
+        # Ensure fields was added with the right defaults
+        assertDefault(date_of_birth_auto_now, 'date_of_birth_auto_now', now.date())
```
Comment (timgraham) [inline tests/schema/tests.py]: @shaib, on Oracle this is failing with:
```
    self.assertEqual(cursor.fetchall()[0][0], expected_default)
AssertionError: datetime.datetime(2000, 1, 1, 0, 0) != datetime.date(2000, 1, 1)
```
I didn't investigate too much but is there something obvious about Oracle datetime handling that would cause this (basically datetime is returned when date is expected)?

**Comment by timgraham** [comment]
@yakky, for future reference, don't forget to uncheck "Patch needs improvement" on the ticket so that it returns to the review queue.

File: tests/schema/tests.py [hunk 19]
```diff
@@ -1832,3 +1833,67 @@ def test_alter_pk_with_self_referential_field(self):
...
+                                                                   mocked_tz):
+        """
+        #25005 - effective_default() should be used for DateField,
+        DateTimeField and TimeField if auto_now or auto_add_now is set.
+        """
+        def assertDefault(field, field_name, expected_default):
+            with connection.cursor() as cursor:
+                editor.add_field(Author, field)
+                cursor.execute("SELECT %s FROM schema_author;" % field_name)
+                self.assertEqual(cursor.fetchall()[0][0], expected_default)
```
Comment (shaib) [inline tests/schema/tests.py]: Even when restricted to tests and defined as an inner function, the name `assert...` for a function with a significant side effect (schema change) is disturbing. So is the use of the surrounding-function variable `editor` (which is done outside of the relevant `with` block). Please restructure this.

File: tests/schema/tests.py (hunk 18 above)
Comment (shaib) [inline tests/schema/tests.py]: Oracle doesn't really have a date-without-time data type. If you're selecting data using raw SQL, you need to account for this yourself.

Comment (yakky) [inline tests/schema/tests.py]: Any way to run these tests somewhere?

File: docs/ref/models/fields.txt (hunk 16 above)
Comment (yakky) [inline docs/ref/models/fields.txt]: @timgraham  Given the changes required for oracle I will edit this myself

File: tests/schema/tests.py (hunk 19 above)
Comment (yakky) [inline tests/schema/tests.py]: @shaib how about ``addFieldAndCheckDefaultValue`` name?

File: tests/schema/tests.py (hunk 18 above)
Comment (shaib) [inline tests/schema/tests.py]: You mean on Oracle? There's Django's CI infrastructure (I think Oracle is run for all PRs automatically, but it may need special invocation). Other than that, you can [set up an environment locally](https://code.djangoproject.com/wiki/OracleTestSetup) but it's a little involved if you only need it for one patch.

File: tests/schema/tests.py (hunk 19 above)
Comment (shaib) [inline tests/schema/tests.py]: I'd prefer something like `check_added_field_default()`; but I'd really rather see it defined as an independent method (or even function), rather than an inner function. It seems like a useful general utility for schema tests.

File: docs/ref/models/fields.txt (hunk 16 above)
Comment (timgraham) [inline docs/ref/models/fields.txt]: Sure, here are the edits I had made locally: http://dpaste.com/3J1M7K1
I think it's fine to squash the commits since adding an untested feature in a commit doesn't make much sense to me.

File: tests/schema/tests.py (hunk 19 above)
Comment (yakky) [inline tests/schema/tests.py]: Added as method and made it more general https://github.com/django/django/pull/6300/files#diff-888b45a0a8f38ee67e8d22403cf994dbR128

File: tests/schema/tests.py (hunk 18 above)
Comment (yakky) [inline tests/schema/tests.py]: Got it how to trigger oracle. Will trigger once these passes on default platforms

File: docs/ref/models/fields.txt (hunk 16 above)
Comment (yakky) [inline docs/ref/models/fields.txt]: Added to this commit and squashed as a single commit

**Comment by yakky** [comment]
buildbot, test on oracle

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
@yakky, do you need help with the Oracle failure or are you still planning to try to investigate and/or fix that?

**Comment by yakky** [comment]
@timgraham I should be able to nail it tomorrow. Zero time to work on this in the past weeks, sorry

**Comment by yakky** [comment]
buildbot, test on oracle.

**Comment by yakky** [comment]
buildbot, test on oracle.

**Comment by yakky** [comment]
@timgraham @shaib oracle fixed!

**Comment by timgraham** [comment]
merged in 49c57f8565d4b655de5161d299a053eddfc00bf2, thanks!

**Comment by yakky** [comment]
Thanks @timgraham
